### PR TITLE
[OB-2824] feat: add retry to web requests

### PR DESCRIPTION
**Pull request by MAG-AW**
connected-spaces-platform Pull Request

**For the reviewer**

* [ ] If required, are the changes covered by appropriate tests?
* [ ] Are any public-facing API changes well documented?
* [ ] Is the code easily readable and extensible and/or follow existing conventions?

### CODE REVIEW

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp [hunk 1]
```diff
@@ -14,11 +14,12 @@
  * limitations under the License.
  */
 
-#include "EmscriptenWebClient.h"
+//#ifdef CSP_WASM
```
Comment (MAG-MichaelK) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:17]: Can we remove this before committing?

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp [hunk 2]
```diff
@@ -14,11 +14,12 @@
...
 
 #include "CSP/Common/Map.h"
 #include "CSP/Common/String.h"
 #include "Debug/Logging.h"
+#include "EmscriptenWebClient.h"
```
Comment (MAG-MichaelK) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:22]: This should be the first include. If clang-format is moving it, put a new line after it.

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp [hunk 3]
```diff
@@ -101,11 +102,26 @@ void OnFetchSuccessOrError(emscripten_fetch_t* Fetch)
...
+{
+	auto* UserData = reinterpret_cast<csp::web::HttpRequest*>(Fetch->userData);
+
+	if (UserData->Retry())
+	{
+		CSP_LOG_WARN_MSG("Retrying failed emscripten request\n");
+	}
+	else
+	{
+		OnFetchSuccessOrError(Fetch);
```
Comment (MAG-MichaelK) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:115]: This function name is now misleading. We should move the success code to a new `OnFetchSuccess` function, and move the error-related code to `OnFetchError`.

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp [hunk 4]
```diff
@@ -101,11 +102,26 @@ void OnFetchSuccessOrError(emscripten_fetch_t* Fetch)
...
+	}
+}
+
 void OnFetchProgress(emscripten_fetch_t* Fetch)
 {
 	if (Fetch->totalBytes)
 	{
-		auto* Request = reinterpret_cast<csp::web::HttpRequest*>(Fetch->userData);
+		auto UserData = Fetch->userData;
```
Comment (MAG-MichaelK) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:123]: This assignment can be removed.

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.h [hunk 5]
```diff
@@ -15,6 +15,10 @@
  */
 #pragma once
 
+#include "CSP/CSPCommon.h"
+
+//#ifdef CSP_WASM
```
Comment (MAG-MichaelK) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.h:20]: Remove please! Also, the header above isn't needed if this symbol check is removed.

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp (hunk 3 above)
Comment (MAG-MichaelK) [reply Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:115]: Happy to leave this one as is if that's the case then.

File: Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp [hunk 6]
```diff
@@ -101,6 +101,20 @@ void OnFetchSuccessOrError(emscripten_fetch_t* Fetch)
 	emscripten_fetch_close(Fetch);
 }
 
+void OnFetchError(emscripten_fetch_t* Fetch)
+{
+	auto* Request = reinterpret_cast<csp::web::HttpRequest*>(Fetch->userData);
+
+	if (Request->Retry())
+	{
+		CSP_LOG_WARN_MSG("Retrying failed emscripten request\n");
```
Comment (MAG-SamBirley) [inline Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:110]: Can we put this on the verbose level? Emitting to regular log level feels like it risks being spammy

Comment (MAG-SamBirley) [reply Library/src/Web/EmscriptenWebClient/EmscriptenWebClient.cpp:110]: No that's perfect, sorry I misread!

**Review by MAG-SamBirley — APPROVED** [review]

**Pull request merged**